### PR TITLE
feat: multi-file ERA5 loading and GCS/CDS comparison (#51)

### LINKED ISSUE #51: ERA5 loader silently clamps out-of-range timestamps to nearest boundary
## Problem

The ERA5 field closures (wind, wave) use `jax.scipy.ndimage.map_coordinates` with `mode="nearest"`. When `t` exceeds the temporal extent of the loaded NetCDF file, the query is silently clamped to the last available timestep.

**Example**: A route departing Dec 28 2024 with a 10-day crossing needs weather data into Jan 7 2025. If only 2024 data is loaded, segments arriving in 2025 silently use Dec 31 2024 18:00 weather — frozen in time, no error, no warning.

## Impact

- Optimization produces routes that ignore actual weather conditions for the later portion of the voyage
- Performance/cost calculations are based on stale data
- The user has no indication this is happening

## Proposed solution

1. **Download script**: Accept a date range (not just a single year) so overlapping data can be fetched:
   ```bash
   uv run scripts/download_era5.py --start 2024-12-01 --end 2025-02-01
   ```
2. **Loader**: Support loading and concatenating multiple NetCDF files along the time axis
3. **Boundary warning**: Log a warning (or raise) when `t` exceeds the dataset's temporal range

## Workaround

For now, download a second year and concatenate manually with xarray, or ensure departures are early enough that the full voyage fits within the downloaded year.

### PR DESCRIPTION
**Pull request by fjsuarez**
## Summary

Addresses #51 — late-year weather data gap caused by single-file ERA5 loading.

### Changes

**Loader** (`routetools/era5/loader.py`):
- `_normalize_time_coord()`: renames `time` → `valid_time` for consistent dimension naming across CDS and GCS sources
- `_load_datasets()`: opens one or more NetCDF files, normalizes time dims, concatenates along time axis
- `load_era5_windfield`, `load_era5_vectorfield`, `load_era5_wavefield`: accept `path: list[Path]` for multi-file concat + optional `voyage_hours` parameter that logs a warning when data coverage is insufficient

**Downloaders**:
- GCS (`download_gcs.py`): `_detect_time_dim()`, `_normalize_time_dim()`, `months` parameter, partial-year filenames
- CDS (`download_cds.py`): partial-year filename support

**CLI** (`scripts/download_era5.py`): `--months` flag forwarded to both backends

**Comparison script** (`scripts/compare_era5_sources.py`): downloads a subset from GCS and compares against existing CDS files, reporting grid alignment, NaN patterns, and value differences

**Dependencies**: added `gcsfs`, `zarr` for GCS access

### GCS vs CDS comparison results (Jan 2024 Atlantic)

| Field | Max abs diff | allclose(atol=1e-3) |
|-------|-------------|---------------------|
| u10   | 4.33e-04    | ✅ True             |
| v10   | 3.92e-04    | ✅ True             |
| swh   | 3.52e-04    | ✅ True             |
| mwd   | 1.06e-02    | ❌ False (0.01° max) |

Grids, timestamps, and NaN patterns are **identical**. Value differences are sub-1e-3 (float32 encoding differences between ARCO-ERA5 Zarr and CDS NetCDF pipelines).

### Tests
- 7 new tests: `valid_time` dim, multi-file concat, mixed dim concat, voyage coverage warning
- 921 total tests pass

Closes #51

### CODE REVIEW

**Comment by daniprec** [comment]
Small follow-up to keep this branch usable on a fresh checkout.

Changes pushed:
- `eb8c229` removes the unconditional `tool.uv.find-links` setting and updates the README to use `uv sync --extra swopp3 --find-links release_package/wheels` only for the optional SWOPP3 wheel install. This avoids `uv run` / `uv sync` / `make hooks` / `make test` failing when `release_package/wheels` is not present locally.
- `5b4972c` keeps `scripts/compare_scorers.py` pre-commit clean so the branch still passes repo validation after the environment fix.

Validation run locally after the change:
- `uv sync`
- `make hooks`
- `make test`

Current result: hooks pass, and tests pass with 236 passed / 711 skipped. The skips are the existing `swopp3_performance_model`-not-installed skips.

**Comment by daniprec** [comment]
I checked `docs/swopp3-era5-pipeline` / PR #59 first. That one is focused on fail-fast behavior when SWOPP3 weather inputs are missing. It does not appear to address the remaining concerns below.

I still see two behavior-level issues in this PR:

1. The main SWOPP3 runner still only wires a single ERA5 file per corridor, so the new multi-file loader capability does not seem to reach the default user-facing workflow.
   - `scripts/swopp3_run.py` still accepts single `Path` inputs for `--wind-path*` / `--wave-path*` and then loads exactly one file per corridor via `load_dataset_epoch(wp)`, `load_era5_windfield(wp)`, and `load_era5_wavefield(wp)`.
   - That means the late-year fix looks incomplete at the CLI level: the loader can concatenate multiple files, but the main SWOPP3 path still cannot pass them through.
   - Given the 2024 departure schedule and 583 h Pacific passages, this still looks like an end-of-data clamp risk for late departures.

2. `scripts/compare_era5_sources.py` still compares mismatched temporal resolutions by default.
   - CDS now defaults to hourly data in `routetools/era5/download_cds.py`.
   - The comparison script still downloads GCS data with `time_step=6`.
   - The comparator treats shape mismatches as failures, so the default comparison path can report false negatives even when the underlying datasets are otherwise aligned.

Smaller note:
- `routetools/era5/download_gcs.py` documents the month-by-month path as bounded-memory, but `_download_by_month()` still does `xr.open_mfdataset(...); ds_merged.load()`, which re-materializes the merged dataset in memory. That may still be too heavy for the full-year hourly case this path is trying to improve.

So my current view is: the PR is in much better shape than before, but I would still want the SWOPP3 runner wiring and the source-comparison default mismatch addressed before giving this a full thumbs-up.

**Comment by daniprec** [comment]
Follow-up on my earlier review: the functional issues I raised are now addressed.

- `3df7eaf` wires the SWOPP3 runner to pick up continuation ERA5 files, so the multi-file loader support now reaches the user-facing path.
- `3713296` makes `compare_era5_sources.py` use hourly GCS downloads by default, matching CDS, and also cleans up dataset handle usage.
- `7524c2b` removes the eager merged-dataset materialization in the month-by-month GCS path.

Validation run locally on this branch:
- `make hooks`
- `make test`

Current result: hooks pass, and tests pass with 243 passed / 711 skipped.

From my side, I no longer see a blocking behavior issue in this PR and would be okay merging it.

**Review by daniprec — APPROVED** [review]
